cumulusci/core/tasks.py

```diff
@@ -226,6 +236,115 @@ def freeze(self, step):
         )
         return [ui_step]
 
+    def set_resumption_file(self, file):
+        pass
+
+
+class ResumableTask(BaseTask):
+    """A task which can be resumed
+
+    Resumable tasks have a different internal API and store
+    their state at a URL location instead of only on the
+    Python stack.
+
+    Attributes of resumable tasks are persisted by default.
+
+    __setattr__ is overriden to require you to declare values and types
+    in order to encourage you to think through what should be
+    persisted and what should not.
+
+    Add transient attributes through self.set_nonpersistent_value to
+    prevent them from being persisted.
+    """
+
+    options: dict = {}
+    name: str = None
+    logger: Any = None
+    resumption_fs_resource: Any = None
+    project_config: dict = None
+    task_config: dict = None
+    org_config: dict = None
+    poll_count: int = None
+    poll_interval_level: int = None
+    poll_interval_s: int = None
+    poll_complete: bool = None
+    return_values: dict = None
+    result: Any = None
+    flow: Any = None
+    stepnum: Any = None
+
+    default_excluded_fields = (
+        "name",
+        "logger",
+        "task_options",
+        "project_config",
+        "task_config",
+        "org_config",
+        "poll_count",
+        "poll_interval_level",
+        "poll_interval_s",
+        "poll_complete",
+        "return_values",
+        "result",
+        "flow",
+        "stepnum",
+        "excluded_fields",
+        "working_path",
+        "sf",
+        "bulk",
+        "tooling",
+    )
+
+    def __init__(self, *args, **kwargs):
+        self.__dict__["excluded_fields"] = set(self.default_excluded_fields)
+
+        if "resumption_fs_resource" in kwargs:
+            self.set_nonpersistent_value(
+                "resumption_fs_resource", kwargs.pop("resumption_fs_resource")
+            )
+
+        super().__init__(*args, **kwargs)
+
+    def _run_task(self):
+        self._start()
+        self._run_steps()
+        self.resumption_fs_resource.cleanup()
+        self._after_finished()
+
+    def _start(self):
+        "Override this: Do any kind of task setup or logging that is necessary."
+        self.logger.info(f"Starting task {self.__class__.name}")
+
+    def _run_steps(self):
+        "Run multiple steps until finished. Usually do not override this."
+        while not self._is_finished():
+            self._run_step()
+            self.resumption_fs_resource.state_data = self.dict()
+            self.resumption_fs_resource.save()
```

Comment:
Won't an interruption in _run_step generally raise an exception that prevents this code from running to save the state?

Reply:
I think that's the right design. Remember we had used the metaphor of a "transaction". Saving after an exception would save some the state halfway through a "transaction" with variables in who knows what state?

I am inspired by the concept of "crash only software." You write the software on the presumption that it could just die anywhere anytime with no warning and then it will automatically handle cases where it dies WITH warning as a side effect.

https://en.wikipedia.org/wiki/Crash-only_software

If you do have state that you want to save instantly, arguably a relational database would be a better place for it and I do think that our tasks should have some form of easy access to an RDB.

Reply:
Got it, I was thinking about this wrong.

It does mean we should be sure to call out some warnings in the docs about writing resumable tasks that do non-idempotent actions. (And for the generate/load task, we should document that resuming it has the potential to insert extra data.)